File: 733. Flood Fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        if newColor == image[sr][sc]:
            return image
        self.dfs(image, sr, sc, newColor, image[sr][sc])
        return image
    
    def dfs(self, image, sr, sc, newColor, oriColor):    
        if 0 > sr or sr >= len(image) or 0 > sc or sc >= len(image[0]):
            return
        if image[sr][sc] == oriColor:
            image[sr][sc] = newColor
            self.dfs(image, sr+1, sc, newColor, oriColor)
            self.dfs(image, sr-1, sc, newColor, oriColor)
            self.dfs(image, sr, sc+1, newColor, oriColor)
            self.dfs(image, sr, sc-1, newColor, oriColor)
        

File: 733. Flood Fill.java
/**
 * BFS solution
 * Time O(n)
 * Space O(n)
 */
class Solution {
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        if (image[sr][sc] == newColor) {
            return image;
        }
        Deque<Integer> dq = new ArrayDeque<>();
        int r = image.length, c = image[0].length;
        int oriColor = image[sr][sc];
        image[sr][sc] = newColor;
        dq.offer(sr*c+sc);
        while (!dq.isEmpty()) {
            int p = dq.poll();
            int pr = p / c, pc = p % c;
            if (pr > 0 && image[pr-1][pc] == oriColor) {
                image[pr-1][pc] = newColor;
                dq.offer((pr-1)*c+pc);
            }
            if (pr < r-1 && image[pr+1][pc] == oriColor) {
                image[pr+1][pc] = newColor;
                dq.offer((pr+1)*c+pc);
            }
            if (pc > 0 && image[pr][pc-1] == oriColor) {
                image[pr][pc-1] = newColor;
                dq.offer(pr*c+pc-1);
            }
            if (pc < c-1 && image[pr][pc+1] == oriColor) {
                image[pr][pc+1] = newColor;
                dq.offer(pr*c+pc+1);
            }
        }
        return image;
    }
}